### PR TITLE
[RFC] Combinded rpc arg

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This PR proposes a way to combine the RPC argument struct with the main struct when using the SMC ABI. This approach is already used with the FF-A ABI. This PR also enables efficient caching of argument structs in the kernel driver, see corresponding changes at https://github.com/jenswi-linaro/linux-1/tree/rpc_arg

### CODE REVIEW

File: core/arch/arm/include/optee_ffa.h [hunk 1]
```diff
@@ -82,8 +82,16 @@
  *                   as the second MSG arg struct for
  *                   OPTEE_FFA_YIELDING_CALL_WITH_ARG.
  *        Bit[31:8]: Reserved (MBZ)
- * w5-w7: Note used (MBZ)
+ * w5:	  Bitfield of secure world capabilities OPTEE_FFA_SEC_CAP_* below,
+ *	  unused bits MBZ.
+ * w6-w7: Not used (MBZ)
+ */
+/*
+ * Secure world supports giving an offset into the argument shared memory
```
Comment (jforissier) [inline core/arch/arm/include/optee_ffa.h:90]: s/giving/receiving/? If I understand correctly, the caller gives the offset and secure world receives it.

File: core/arch/arm/kernel/thread_optee_smc.c [hunk 2]
```diff
@@ -97,108 +97,194 @@ static void thread_rpc_free_arg(uint64_t cookie)
...
 
+static uint32_t std_smc_entry(uint32_t a0, uint32_t a1, uint32_t a2,
+			      uint32_t a3 __unused)
+{
+	const bool with_rpc_arg = true;
+
+	switch (a0) {
+	case OPTEE_SMC_CALL_WITH_ARG:
+		return std_entry_with_parg(reg_pair_to_64(a1, a2),
+					   !with_rpc_arg);
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_optee_smc.c:276]: Nice :wink:

File: core/arch/arm/include/optee_ffa.h [hunk 3]
```diff
@@ -113,6 +121,8 @@
  *	  OPTEE_MSG_GET_ARG_SIZE(num_params) follows a struct optee_msg_arg
  *	  for RPC, this struct has reserved space for the number of RPC
  *	  parameters as returned by OPTEE_FFA_EXCHANGE_CAPABILITIES.
+ *	  MBZ unless the bit OPTEE_FFA_SEC_CAP_ARG_OFFSET is received with
+ *	  OPTEE_FFA_EXCHANGE_CAPABILITIES.
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/include/optee_ffa.h:125]: Alright

**Review by jforissier — COMMENTED** [review]
- "core: provide dummy mobj_reg_shm_get_by_cookie()":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

- "core: add OPTEE_SMC_CALL_WITH_RPC_ARG" and "core: add OPTEE_FFA_SEC_CAP_ARG_OFFSET":
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Tags applied

**Comment by jforissier** [comment]
@jbech-linaro re. the IBART error. I am suspecting a dependency issue (possibly revealed by the recently introduced [parallel build](https://github.com/OP-TEE/build/commit/e9bff1c86e93820616e5283615d32683ac0c6d03)):
```
  9117:  /home/optee/devel/hikey/out-br/build/busybox-1.34.1/scripts/gcc-version.sh: line 11: /home/optee/devel/hikey/out-br/per-package/busybox/host/bin/aarch64-linux-gnu-gcc: No such file or directory
```
I am restarting the build and we'll see.

**Comment by jenswi-linaro** [comment]
> Yes, and even so, there is still one month to the next release which is when ABI changes really matter.

Yes, we should be know for sure by then.

**Pull request merged**